src/source-dpdk.c

```diff
@@ -0,0 +1,494 @@
+                TmqhOutputPacketpool(ptv->tv, p);
+                DPDKFreeMbufArray(ptv->received_mbufs, nb_rx - i - 1, i + 1);
+                SCReturnInt(EXIT_FAILURE);
+            }
+
+            if (TmThreadsSlotProcessPkt(ptv->tv, ptv->slot, p) != TM_ECODE_OK) {
+                TmqhOutputPacketpool(ptv->tv, p);
+                DPDKFreeMbufArray(ptv->received_mbufs, nb_rx - i - 1, i + 1);
+                SCReturnInt(EXIT_FAILURE);
+            }
+        }
+
+        /* Trigger one dump of stats every second */
+        current_time = DPDKGetSeconds();
+        if (current_time != last_dump) {
+            DPDKDumpCounters(ptv);
+            last_dump = current_time;
+        }
+    }
+
+    SCReturnInt(TM_ECODE_OK);
+}
+
+/**
+ * \brief Init function for ReceiveDPDK.
+ *
+ * \param tv pointer to ThreadVars
+ * \param initdata pointer to the interface passed from the user
+ * \param data pointer gets populated with DPDKThreadVars
+ *
+ */
+TmEcode ReceiveDPDKThreadInit(ThreadVars *tv, const void *initdata, void **data)
+{
+    SCEnter();
+    int retval;
+    DPDKIfaceConfig *dpdk_config = (DPDKIfaceConfig *)initdata;
+
+    if (initdata == NULL) {
+        SCLogError(SC_ERR_INVALID_ARGUMENT, "DPDK Configuration failed");
+        dpdk_config->DerefFunc(dpdk_config);
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    DPDKThreadVars *ptv = SCCalloc(1, sizeof(DPDKThreadVars));
+    if (unlikely(ptv == NULL)) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Unable to allocate memory");
+        dpdk_config->DerefFunc(dpdk_config);
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    ptv->tv = tv;
+    ptv->pkts = 0;
+    ptv->bytes = 0;
+    ptv->livedev = LiveGetDevice(dpdk_config->iface);
+
+#ifdef PACKET_STATISTICS
+    ptv->capture_dpdk_packets = StatsRegisterCounter("capture.packets", ptv->tv);
+    ptv->capture_dpdk_drops = StatsRegisterCounter("capture.drops", ptv->tv);
+#endif
+
+    ptv->copy_mode = dpdk_config->copy_mode;
+    ptv->checksum_mode = dpdk_config->checksum_mode;
+
+    ptv->threads = dpdk_config->threads;
+    ptv->port_id = dpdk_config->port_id;
+    ptv->out_port_id = dpdk_config->out_port_id;
+    uint16_t queue_id = SC_ATOMIC_ADD(dpdk_config->queue_id, 1);
+    ptv->queue_id = queue_id;
+    // pass the pointer to the mempool and then forget about it. Mempool is freed in thread deinit.
+    ptv->pkt_mempool = dpdk_config->pkt_mempool_array[queue_id];
+    dpdk_config->pkt_mempool_array[queue_id] = NULL;
+
+    // the last thread starts the device
+    if (queue_id == dpdk_config->threads - 1) {
+        retval = rte_eth_dev_start(ptv->port_id);
+        if (retval < 0) {
+            SCLogError(SC_ERR_DPDK_INIT, "Error (err=%d) during device startup of port %u", retval,
+                    ptv->port_id);
+            dpdk_config->DerefFunc(dpdk_config);
+            SCReturnInt(TM_ECODE_FAILED);
+        }
+
+        retval = dpdk_config->flags & DPDK_PROMISC ? rte_eth_promiscuous_enable(ptv->port_id)
```

Comment:
any specific reason to enable or disable promiscuous mode after start. Please note this is not a clean usage because RX-DMA will happen once start is invoked. So this breaks the functionality. Rework

Reply:
you miss the point, please recheck any of the examples when does actual RX of packets happen. once you understand rework this correctly.